Mathlib/RingTheory/Subsemiring/Basic.lean

```diff
@@ -0,0 +1,1429 @@
+/-
+Copyright (c) 2020 Yury Kudryashov All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
+
+! This file was ported from Lean 3 source module ring_theory.subsemiring.basic
+! leanprover-community/mathlib commit f93c11933efbc3c2f0299e47b8ff83e9b539cbf6
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Algebra.Module.Basic
+import Mathlib.Algebra.Ring.Equiv
+import Mathlib.Algebra.Ring.Prod
+import Mathlib.Algebra.Order.Ring.InjSurj
+import Mathlib.Algebra.GroupRingAction.Subobjects
+import Mathlib.Data.Set.Finite
+import Mathlib.GroupTheory.Submonoid.Centralizer
+import Mathlib.GroupTheory.Submonoid.Membership
+
+/-!
+# Bundled subsemirings
+
+We define bundled subsemirings and some standard constructions: `complete_lattice` structure,
+`subtype` and `inclusion` ring homomorphisms, subsemiring `map`, `comap` and range (`srange`) of
+a `ring_hom` etc.
+-/
+
+
+-- Porting note: currently global `open BigOperators`
+
+universe u v w
+
+section AddSubmonoidWithOneClass
+
+/-- `add_submonoid_with_one_class S R` says `S` is a type of subsets `s ≤ R` that contain `0`, `1`,
+and are closed under `(+)` -/
+class AddSubmonoidWithOneClass (S : Type _) (R : outParam <| Type _) [AddMonoidWithOne R]
```

Comment:
`R` shouldn't be an `outParam` if it's already an `outParam` for `SetLike`, following #1832.
```suggestion
class AddSubmonoidWithOneClass (S R : Type _) [AddMonoidWithOne R]
```